src/compiler/checker.ts

```diff
@@ -12125,8 +12125,14 @@ module ts {
             if (!parameter.type) {
                 return grammarErrorOnNode(parameter.name, Diagnostics.An_index_signature_parameter_must_have_a_type_annotation);
             }
-            if (parameter.type.kind !== SyntaxKind.StringKeyword && parameter.type.kind !== SyntaxKind.NumberKeyword) {
-                return grammarErrorOnNode(parameter.name, Diagnostics.An_index_signature_parameter_type_must_be_string_or_number);
+            if (parameter.type.kind === SyntaxKind.TypeReference) {

```

Comment:
Since you do the same thing for this and the next `if` branches, factor these checks out into a helper function named `isValidTypeOfIndexSignature`.

Do the check for `StringKeyword` and `NumberKeyword` first so it bails out early.